pkg/migrator/migrator.go

```diff
@@ -0,0 +1,461 @@
+			if err != nil {
+				return errors.Wrapf(err, "deleting the CRD registration for \"%s\"", tprKind)
+			}
+		}
+
+		if ms.tprRegDel {
+			_, err := m.tprClient.Create(tpr)
+			if err != nil {
+				return errors.Wrapf(err, "recreating the TPR registration for \"%s\"", tprKind)
+			}
+		}
+
+		if ms.tprDataDel {
+			err := m.restoreFromCM(tprKind, tprName, configmapName)
+			if err != nil {
+				return errors.Wrapf(
+					err,
+					"recreating the TPR data for \"%s\" from config-map \"%s\"",
+					tprKind, configmapName)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *Migrator) moveTPRDataToCM(tprName, configmapName string) error {
+	if m.cmSuffix != "" {
+		configmapName = configmapName + "-" + m.cmSuffix
+	}
+
+	req := m.restClient.Get().
+		Namespace(api.NamespaceAll).
+		Resource(tprName).
+		FieldsSelectorParam(nil)
+
+	tprdata, err := req.DoRaw()
+	if err != nil {
+		return nil
+	}
+
+	cm := v1.ConfigMap{}
+
+	cm.ObjectMeta.SetNamespace("default")
+	cm.ObjectMeta.SetName(configmapName)
+
+	cm.Data = map[string]string{"backup": string(tprdata)}
+
+	_, err = m.cmClient.Create(&cm)
+	return err
+}
+
+func (m *Migrator) deleteTPR(tprName string) error {
+	return m.tprClient.Delete(tprName, &metav1.DeleteOptions{})
+}
+
+func (m *Migrator) deleteCRD(crdName string) error {
+	return m.crdClient.Delete(crdName, &metav1.DeleteOptions{})
+}
+
+func (m *Migrator) createCRD(crd *extensionsobj.CustomResourceDefinition) error {
+	_, err := m.crdClient.Create(crd)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return errors.Wrapf(err, "Creating CRD: %s", crd.Spec.Names.Kind)
+	}
+
+	return nil
+}
+
+func (m *Migrator) restoreFromCM(tprKind, tprName, configmapName string) error {
+	if m.cmSuffix != "" {
+		configmapName = configmapName + "-" + m.cmSuffix
+	}
+
+	cm, err := m.cmClient.Get(configmapName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "creating config map client")
+	}
+
+	tprdata := cm.Data["backup"]
+
+	us := &unstructured.UnstructuredList{}
+	err = json.Unmarshal([]byte(tprdata), us)
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling into UnstructuredList")
+	}
+
+	for _, item := range us.Items {
+		item.SetResourceVersion("") // TODO(gouthamve): HACK?? Best practices
+		// Namespace issues
+		resClient := m.dynClient.Resource(
+			&metav1.APIResource{
+				Kind:       tprKind,
+				Name:       tprName,
+				Namespaced: true,
+			},
+			item.GetNamespace(),
+		)
+		_, err = resClient.Create(item)
+		if err != nil {
+			return errors.Wrap(err, "restoring tprdata")
+		}
+	}
+
+	return nil
+}
+
+func (m *Migrator) deleteTPRData(tprKind, tprName string) error {
+	nsList, err := m.nsClient.List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "listing namspaces")
+	}
+
+	for _, ns := range nsList.Items {
+		resClient := m.dynClient.Resource(
+			&metav1.APIResource{
+				Kind:       tprKind,
+				Name:       tprName,
+				Namespaced: true,
+			},
+			ns.ObjectMeta.Name,
```

Comment:
Got it. Makes sense. I was just hoping to not require the additional RBAC role for namespaces, but sounds reasonable.